plugins/experimental/icap/icap_plugin.cc

```diff
@@ -0,0 +1,1267 @@
+    data->icap_vc  = NULL;
+    data->icap_vio = NULL;
+  }
+  TSStatIntIncrement(total_icap_invalid, 1);
+  TSDebug(PLUGIN_NAME, "\n%s\n", data->icap_header.c_str());
+  data->err_msg = std::string(msg);
+  /* Signal the upstream vconn if still exists to stop sending data */
+  TSVIO write_vio = TSVConnWriteVIOGet(contp);
+  if (TSVIOBufferGet(write_vio)) {
+    TSContCall(TSVIOContGet(write_vio), TS_EVENT_VCONN_WRITE_COMPLETE, write_vio);
+  }
+
+  TSMBuffer bufp;
+  TSMLoc hdr_loc;
+
+  if (TSHttpTxnTransformRespGet(data->txn, &bufp, &hdr_loc) != TS_SUCCESS) {
+    TSError("[%s] Couldn't retrieve transform response header", PLUGIN_NAME);
+    return 0;
+  }
+  /* Clear all headers from the transform response */
+  if (TSMimeHdrFieldsClear(bufp, hdr_loc) == TS_ERROR) {
+    TSError("[%s] Couldn't clear client response header", PLUGIN_NAME);
+    return 0;
+  }
+  TSHttpHdrStatusSet(bufp, hdr_loc, TS_HTTP_STATUS_BAD_GATEWAY);
+  TSHttpHdrReasonSet(bufp, hdr_loc, TSHttpHdrReasonLookup(TS_HTTP_STATUS_BAD_GATEWAY),
+                     strlen(TSHttpHdrReasonLookup(TS_HTTP_STATUS_BAD_GATEWAY)));
+  TSHandleMLocRelease(bufp, TS_NULL_MLOC, hdr_loc);
+  transform_send_error_msg(contp, data);
+
+  return 0;
+}
+
+/*
+ * handle_icap_headers
+ * Description: This is a good place to determine what to do next based on
+ *              icap eaders from response of icap server.
+ */
+static int
+handle_icap_headers(TSCont contp, TransformData *data)
+{
+  int64_t pos                  = data->icap_header.find("\r\n");
+  std::string icap_status_line = pos != (int64_t)std::string::npos ? data->icap_header.substr(0, pos) : data->icap_header;
+  /* Check icap header to determine whether the scan passed or not */
+  if (data->icap_header.find("X-Infection-Found") != std::string::npos ||
+      data->icap_header.find("X-Violations-Found") != std::string::npos) {
+    TSStatIntIncrement(scan_failed, 1);
+  } else {
+    TSStatIntIncrement(scan_passed, 1);
+  }
+  /* If debug-mode is enabled, add header to log ICAP status */
+  if (debug_enabled) {
+    if (icap_status_line.find("506") != std::string::npos) {
+      setup_icap_status_header(data, "@ICAP-Status", "ICAP server is too busy");
+      TSDebug(PLUGIN_NAME, "Sending OS response body.");
+      return 1;
+    }
+  }
+
+  return 0;
+}
+
+/*
+ * handle_icap_http_header
+ * Description: This is a good place to determine what to do next based on
+ *              modified http headers from response of icap server.
+ */
+static void
+handle_icap_http_header(TransformData *data)
+{
+  // TSDebug(PLUGIN_NAME, "Handling http header");
+  int64_t pos                  = data->http_header.find("\r\n");
+  std::string http_status_line = pos != (int64_t)std::string::npos ? data->http_header.substr(0, pos) : data->http_header;
+  /* find content length from header if any */
+  std::smatch sm;
+  std::regex e("(Content-Length: )([[:digit:]]+)");
+  regex_search(data->http_header, sm, e);
```

Comment:
This looks like it's high-frequency code, so seems like it would be worth the small effort to avoid regular expression matching:  http://coliru.stacked-crooked.com/a/c65bc52881070179 .